Wrappers/Python/cil/optimisation/algorithms/sampler.py

```diff
@@ -0,0 +1,461 @@
+            The sampler will select from a list of integers {0, 1, …, S-1} with S=num_subsets. 
+
+        sampling_type:str
+            The sampling type used. 
+
+        order: list of integers
+            The list of integers the method selects from using next. 
+
+        shuffle= bool, default=False
+            If True, after each num_subsets calls of next, the sampling order is shuffled randomly. 
+
+        prob: list of floats of length num_subsets that sum to 1. 
+            For random sampling with replacement, this is the probability for each integer to be called by next. 
+
+        seed:int, default=None
+            Random seed for the methods that use a random number generator. If set to None, the seed will be set using the current time.  
+        """
+        self.type = sampling_type
+        self.num_subsets = num_subsets
+        if seed is not None:
+            self.seed = seed
+        else:
+            self.seed = int(time.time())
+        self.generator = np.random.RandomState(self.seed)
+        self.order = order
+        if order is not None:
+            self.iterator = self._next_order
+        self.shuffle = shuffle
+        if self.type == 'random_without_replacement' and self.shuffle == False:
+            self.order = self.generator.permutation(self.order)
+        self.initial_order = self.order
+        self.prob = prob
+        if prob is not None:
+            self.iterator = self._next_prob
+        self.last_subset = self.num_subsets-1
+
+    def _next_order(self):
+        """ 
+        The user should call sampler.next() or next(sampler) rather than use this function. 
+
+        A function of the sampler that selects from a list of integers {0, 1, …, S-1}, with S=num_subsets, the next sample according to the type of sampling.
+
+        This function is used by samplers that sample without replacement. 
+
+        """
+      #  print(self.last_subset)
+        if self.shuffle == True and self.last_subset == self.num_subsets-1:
+            self.order = self.generator.permutation(self.order)
+            # print(self.order)
+        self.last_subset = (self.last_subset+1) % self.num_subsets
+        return (self.order[self.last_subset])
+
+    def _next_prob(self):
+        """ 
+        The user should call sampler.next() or next(sampler) rather than use this function. 
+
+        A function of the sampler that selects from a list of integers {0, 1, …, S-1}, with S=num_subsets, the next sample according to the type of sampling.
+
+        This function us used by samplers that select from a list of integers {0, 1, …, S-1}, with S=num_subsets, randomly with replacement. 
+
+        """
+        return int(self.generator.choice(self.num_subsets, 1, p=self.prob))
+
+    def next(self):
+        """ A function of the sampler that selects from a list of integers {0, 1, …, S-1}, with S=num_subsets, the next sample according to the type of sampling. """
+
+        return (self.iterator())
```

Comment:
self.iterator is always configured because my methods either all hold an order in memory or call randomly with replacement based on a probability. Would there be any sampler that wouldn't do either of these?

Reply:
Perhaps we should have a think about the custom sampler for mantid imaging and this would help consider this class from a user's perspective?